marshal/tests/serialization/set.go

```diff
@@ -0,0 +1,121 @@
+package serialization
+
+import (
+	"bytes"
+	"fmt"
+	"runtime/debug"
+	"testing"
+
+	"github.com/gocql/gocql/marshal/tests/funcs"
+	"github.com/gocql/gocql/marshal/tests/mod"
+	"github.com/gocql/gocql/marshal/tests/utils"
+)
+
+type Sets []*Set
+
+// Set is a tool for generating test cases of marshal and unmarshall funcs.
+// For cases when the function should no error,
+// marshaled data from Set.Values should be equal with Set.Data,
+// unmarshalled value from Set.Data should be equal with Set.Values.
+type Set struct {
+	Data   []byte
+	Values []interface{}
+
+	IssueMarshal   string
+	IssueUnmarshal string
+}
+
+func (s Set) AddModified(mods ...mod.Mod) Set {
+	out := s
+	for _, m := range mods {
+		out.Values = append(out.Values, m.Apply(s.Values)...)
+	}
+	return out
+}
+
+func (s Set) Run(name string, t *testing.T, marshal func(interface{}) ([]byte, error), unmarshal func([]byte, interface{}) error) {
+	t.Logf("test set %s started", name)
+	for i := range s.Values {
+		val := s.Values[i]
```

Comment:
I suggest applying mods directly to values, not to sets. 
Instead of this:
```
	serialization.Set{
		Data:   []byte("\x00\x00"),
		Values: []interface{}{"0"},
	}.AddModified(mod.IntoRef).Run("[0000]str", t, marshal, unmarshal)
```
make this
```
	serialization.Set{
		Data:   []byte("\x00\x00"),
		Values: mod.IntoRef.Append("0"),
	}.Run("[0000]str", t, marshal, unmarshal)
```
In my opinion, it will be easier and clearer this way. And Set.Run function become easier and clearer.

Reply:
> ```
> 	serialization.Set{
> 		Data:   []byte("\x00\x00"),
> 		Values: mod.IntoRef.Append("0"),
> 	}.Run("[0000]str", t, marshal, unmarshal)
> ```
> 
> In my opinion, it will be easier and clearer this way. And Set.Run function become easier and clearer.

Sure, let's do this then:
```
 	serialization.Set{
 		Data:   []byte("\x00\x00"),
 		Values: mod.AddVariants([]interface{}{"0"}, mod.IntoRef),
 	}.Run("[0000]str", t, marshal, unmarshal)
```